pytential/symbolic/execution.py

```diff
@@ -124,31 +162,38 @@ def map_min(self, expr):
                 expr)
 
     def map_node_sum(self, expr):
-        # FIXME: make less CL specific
-        queue = self.array_context.queue
-        return sum(
-                cl.array.sum(grp_ary, queue=queue).get()[()]
-                for grp_ary in self.rec(expr.operand))
+        actx = self.array_context
+        result = sum(actx.np.sum(grp_ary) for grp_ary in self.rec(expr.operand))
+        if not actx._force_device_scalars:
+            result = actx.to_numpy(result)[()]
+
+        return result
 
     def map_node_max(self, expr):
-        # FIXME: make less CL specific
-        queue = self.array_context.queue
-        return max(
-                cl.array.max(grp_ary, queue=queue).get()[()]
-                for grp_ary in self.rec(expr.operand))
+        actx = self.array_context
+        result = max(actx.np.max(grp_ary) for grp_ary in self.rec(expr.operand))
```

Comment:
TIL: `min`, `max` work on device scalars. I imagine it's this:
https://github.com/inducer/pyopencl/blob/a3d0b5cf4aceafa849e23486ace62962d5a2543f/pyopencl/array.py#L1467-L1472
that's doing that.

Reply:
Hm, that forces a bunch of `.get()`. Should it use `reduce(actx.np.maximum, ...)` instead?

Reply:
It does. It's not ideal, but I can live with it for now. It's easy to fix later on when needed.